Mathlib/Analysis/NormedSpace/Star/Basic.lean

```diff
@@ -149,6 +153,25 @@ theorem mul_star_self_ne_zero_iff (x : E) : x * x⋆ ≠ 0 ↔ x ≠ 0 := by
   simp only [Ne, mul_star_self_eq_zero_iff]
 #align cstar_ring.mul_star_self_ne_zero_iff CstarRing.mul_star_self_ne_zero_iff
 
+theorem norm_eq_sSup_norm_mul [Module ℝ E] [IsScalarTower ℝ E E] [BoundedSMul ℝ E] (x : E) :
+    ‖x‖ = sSup { c | ∃ y, c = ‖y * x‖ ∧ ‖y‖ ≤ 1 } := by
+  let x' := ‖x⋆‖⁻¹ • x⋆
+  have h₁ : ‖x‖ = ‖x' * x‖ := by
+    simp [x', smul_mul_assoc, norm_smul, norm_inv, norm_norm, norm_star_mul_self,
+          ← mul_assoc, norm_star]
+  have h₁' : ‖x'‖ ≤ 1 := by
+    by_cases htriv : x = 0
+    · simp [x', htriv]
+    · simp [x', norm_smul, inv_mul_cancel (norm_ne_zero_iff.mpr htriv)]
+  apply Eq.symm
+  refine csSup_eq_of_forall_le_of_forall_lt_exists_gt ⟨‖x‖, x', ⟨h₁, h₁'⟩⟩ ?_ ?_
+  · rintro c ⟨y, hy₁, hy₂⟩
+    rw [hy₁]
+    calc _ ≤ ‖y‖ * ‖x‖ := norm_mul_le _ _
+      _ ≤ 1 * ‖x‖ := by gcongr
+      _ = ‖x‖ := by simp
+  · exact fun c hc => ⟨‖x‖, ⟨x', h₁, h₁'⟩, hc⟩
```

Comment:
So, [IsGreatest.csSup_eq](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/ConditionallyCompleteLattice/Basic.html#IsGreatest.csSup_eq) is a pretty nice result for when the supremum is attained like this. However, there are a few lemmas we're missing to make things really nice:
```
lemma _root_.Set.inter_singleton {α : Type*} (s : Set α) (a : α) :
    s ∩ {a} = {a} ∨ s ∩ {a} = ∅ := by
  simpa using Classical.em (a ∈ s)

lemma _root_.Set.singleton_inter {α : Type*} (s : Set α) (a : α) :
    s ∩ {a} = {a} ∨ s ∩ {a} = ∅ := by
  simpa using Classical.em (a ∈ s)

lemma _root_.norm_mul_le_norm_right {α : Type*} [NonUnitalNormedRing α]
    {a : α} (ha : ‖a‖ ≤ 1) (b : α) : ‖a * b‖ ≤ ‖b‖ := calc
  ‖a * b‖ ≤ ‖a‖ * ‖b‖ := norm_mul_le _ _
  _ ≤ 1 * ‖b‖ := by gcongr
  _ = ‖b‖ := one_mul _

lemma _root_.norm_mul_le_norm_left {α : Type*} [NonUnitalNormedRing α]
    {b : α} (hb : ‖b‖ ≤ 1) (a : α) : ‖a * b‖ ≤ ‖a‖ := calc
  ‖a * b‖ ≤ ‖a‖ * ‖b‖ := norm_mul_le _ _
  _ ≤ ‖a‖ * 1 := by gcongr
  _ = ‖a‖ := mul_one _
```
With those, we get:
```suggestion
    ‖x‖ = sSup { ‖y * x‖ | (y : E) (hy : ‖y‖ ≤ 1) } := by
  by_cases htriv : x = 0
  · obtain (h | h) := Set.inter_singleton {a : ℝ | ∃ y : E, ‖y‖ ≤ 1} 0
    all_goals simp [htriv, Set.setOf_and, h]
  · rw [← norm_eq_zero'] at htriv
    refine Eq.symm <| IsGreatest.csSup_eq ⟨⟨‖x⋆‖⁻¹ • x⋆, ⟨?_, ?_⟩⟩, ?_⟩
    · simp [norm_smul, inv_mul_cancel htriv]
    · simp [smul_mul_assoc, norm_smul, CstarRing.norm_star_mul_self, inv_mul_cancel_left₀ htriv]
    · rintro - ⟨y, hy, rfl⟩
      exact norm_mul_le_norm_right hy x
```